### PR TITLE
Enable OAuth on Yahoo & AOL

### PR DESCRIPTION
**Pull request by nvsnvikram**

### CODE REVIEW

**Comment by Sancus** [comment]
@mkmelin You were working on this right? Is this good to go?

**Comment by nvsnvikram** [comment]
@benbucksch , can you review?

**Comment by mkmelin** [comment]
Haven't had time to do any testing yet unfortunately, but after https://bugzilla.mozilla.org/show_bug.cgi?id=1293958 lands it should work according to nvsnvikram, who I think is working with Yahoo to make this happen.

**Comment by nvsnvikram** [comment]
All the OAuth setup is complete and I have been able to setup Yahoo & AOL accounts on thunderbird. I am waiting for https://bugzilla.mozilla.org/show_bug.cgi?id=1293958 to be merged.

**Comment by Sancus** [comment]
OK, so do you want this to go live now, or should I hold off until the patch above gets merged and makes it into a released version? I don't know if Yahoo/AOL auth is working on current versions or not, or if this will affect it. If it's just broken now then it doesn't matter too much.

File: yahoo.com [hunk 1]
```diff
@@ -23,30 +22,36 @@
...
       <username>%EMAILADDRESS%</username>
-      <usernameField id="login-username" />
-      <passwordField id="login-passwd" />
-      <loginButton id="login-signin" />
+      <usernameField id="login-username"/>
+      <passwordField id="login-passwd"/>
+      <loginButton id="login-signin"/>
     </loginPageInfo>
   </webMail>
+  <documentation url="https://help.yahoo.com/kb/new-mail-for-desktop">
```
Comment (benbucksch) [inline yahoo.com]: This is just the general help site. Please link to the specific pages for Thunderbird, e.g.
https://help.yahoo.com/kb/new-mail-for-desktop/imap-server-settings-yahoo-mail-sln4075.html?impressions=true
and
https://help.yahoo.com/kb/new-mail-for-desktop/pop-access-settings-instructions-yahoo-mail-sln4724.html?impressions=true

File: yahoo.com [hunk 2]
```diff
@@ -23,30 +22,36 @@
       <port>993</port>
       <socketType>SSL</socketType>
       <username>%EMAILADDRESS%</username>
+      <authentication>OAuth2</authentication>
```
Comment (benbucksch) [inline yahoo.com]: Is OAuth2 required yet?
Right now, do end users have to manually change settings (that is, either enable "insecure apps" manually, or create a specific IMAP password) in order to use Yahoo mail in Thunderbird?

File: yahoo.com [hunk 3]
```diff
@@ -23,30 +22,36 @@
...
       <port>465</port>
       <socketType>SSL</socketType>
       <username>%EMAILADDRESS%</username>
+      <authentication>OAuth2</authentication>
       <authentication>password-cleartext</authentication>
     </outgoingServer>
   </emailProvider>
   <webMail>
-    <loginPage url="https://mail.yahoo.com" />
+    <loginPage url="https://mail.yahoo.com"/>
```
Comment (benbucksch) [inline yahoo.com]: Please avoid unnecessary changes.

**Comment by benbucksch** [comment]
@Sancus : For this to merge, the first
1. Thunderbird patch has to merge that makes in Thunderbird support Yahoo OAuth2 and it works in practice
2. It needs to be tested that it actually works, in all cases
3. It has to be released as stable version of Thunderbird to end users
4. all users (or the vast majority, i.e. 98% or so) need to have upgraded to that new version
Only then can we push this config change out to the ISPDB.

File: yahoo.com (hunk 1 above)
Comment (benbucksch) [reply yahoo.com]: (FWIW, pay attention to escape "?" in XML.)

**Comment by Sancus** [comment]
@benbucksch Thanks, that's exactly what I was asking about. I'll leave this to your review since you know more about this than me :)

File: yahoo.com (hunk 2 above)
Comment (nvsnvikram) [reply yahoo.com]: I didn't understand this question. Yahoo! users on OAuth2 enabled clients can: (a) use OAuth2 for authenticating, (b) Use an app or IMAP specific password, or (c) Enable "insecure" apps and use their account password.

**Comment by mkmelin** [comment]
I think 98% is not a reasonable goal. That would take many years if at all. 
But we should get the patch into tb60 and once that has enough uptake it should be ok.

I don't think it's that important to be able to set up new accounts (without tinkering) on an out of date version of the software. Especially as very few regular users will figure out they have to go to the website and select the "allow insecure apps" option to set up an account in Thunderbird at the moment. 

It would also be good to make Thunderbird smarter and don't allow it to select "Oauth2" if the host is not listed in OAuth2Providers.jsm. (Didn't verify if that's the case currently.)

**Comment by nvsnvikram** [comment]
I have amended the PR with review comments: (a) Documentation url now points to the IMAP & POP page, and (b) removed the space cleanup in unrelated code. This should be good to go whenever we merge the PR.

**Comment by benbucksch** [comment]
> It would also be good to make Thunderbird smarter and don't allow it to select "Oauth2"
> if the host is not listed in OAuth2Providers.jsm.

That is already the case

File: yahoo.com [hunk 4]
```diff
@@ -23,22 +22,28 @@
...
     </incomingServer>
     <outgoingServer type="smtp">
       <hostname>smtp.mail.yahoo.com</hostname>
       <port>465</port>
       <socketType>SSL</socketType>
       <username>%EMAILADDRESS%</username>
+      <authentication>OAuth2</authentication>
       <authentication>password-cleartext</authentication>
     </outgoingServer>
+    <documentation url="https://help.yahoo.com/kb/new-mail-for-desktop#topic=TOP_POPIMAP">
```
Comment (benbucksch) [inline yahoo.com]: That is again a search result page. Please list the exact pages, i.e.
https://help.yahoo.com/kb/new-mail-for-desktop/pop-access-settings-instructions-yahoo-mail-sln4724.html
and
https://help.yahoo.com/kb/new-mail-for-desktop/imap-server-settings-yahoo-mail-sln4075.html
You can have several `<documentation>` nodes.

Comment (benbucksch) [reply yahoo.com]: @nvsnvikram : Thanks for the other changes.

Comment (nvsnvikram) [reply yahoo.com]: added imap and pop documentation urls

**Comment by nvsnvikram** [comment]
I am ok with whenever you folks decide to merge. Looks like older versions of Thunderbird ignore OAuth2 in the configuration if host is not listed in OAuth2Providers.jsm and switches back to username/password. Why do we need to wait for 85% migration if there is no problem with user experience?

**Comment by benbucksch** [comment]
> Looks like older versions of Thunderbird ignore OAuth2 in the configuration if host is not listed in OAuth2Providers.jsm

OK, great, I didn't know that for sure. In that case, I would merely wait for a general stable release of a TB with the patch, not waiting for adaption.

**Comment by nvsnvikram** [comment]
Can we merge this now that Thunderbird 60 is released: https://www.thunderbird.net/en-US/thunderbird/60.0/releasenotes/ ?

**Comment by mkmelin** [comment]
Auto updates from 52 series are now finally available, so let's get this in.

**Comment by benbucksch** [comment]
Auto-updates are on a throttle right now.
We need to wait until most users have updated.

**Comment by mkmelin** [comment]
The 52 series just ignore this entry so it doesn't hurt there. Going forward, new users would be on 60 anyway, and may want to use it.

**Comment by benbucksch** [comment]
FWIW, what this change means:
* Without this change (until how), you can use Thunderbird with Yahoo and log in with password, but you have to enable "less secure apps" in the Yahoo account settings. Once you do that, Thunderbird can use standard password login, save your password. Yahoo IMAP works like any standard IMAP. You do not have to enter your password ever again, until you change it.
* With this change, a webpage pops up and the user has to enter the email address and password at a Yahoo page. Advantage is that the end users does not have to find and enable "less secure apps". Another advantage is that this works with 2 factor auth. Disadvantage is that the user has to manually log in. The session expiration time can be set arbitrarily by Yahoo. So, the log in page might appear at random times.

At least that's how I understand the situation.

**Comment by benbucksch** [comment]
> The 52 series just ignore this entry so it doesn't hurt there

Have you actually tried it? From what I understand, 52 supports OAuth2, but only for Google. I don't know whether the fallback is graceful or breaks. Has anybody tested TB 52 with this particular config file and a Yahoo account? I haven't.

**Comment by benbucksch** [comment]
See my comment from Jul 19:
> It needs to be tested that it actually works, in all cases

Nobody confirmed that this was actually tested, and under which circumstances.

**Comment by mkmelin** [comment]
Testing completely live will be possible soon by setting the mailnews.auto_config_url pref to https://autoconfig-stage.thunderbird.net/v1.1/. When tested we can merge to the prod branch to make it live.

**Comment by benbucksch** [comment]
Manual testing is possible by copying the XML file to the TB install dir  _your-local-thunderbord-install-dir_/isp/yahoo.com.xml on the harddisk , see [docs](https://developer.mozilla.org/en-US/docs/Mozilla/Thunderbird/Autoconfiguration) .

**Comment by mkmelin** [comment]
I tested now with autoconfig-stage: 
For 52, finds the settings but can't be set up (complains about username/password), until I select manual config and change OAuth2 -> Normal password. For 60, works as expected.

So, we should wait until uptake of 60 is more complete until we merge this to prod.

**Comment by benbucksch** [comment]
> I tested now

Thanks

> For 52, finds the settings but can't be set up (complains about username/password), until I select manual config and change OAuth2 -> Normal password

OK, too bad. For me, that counts as "breaking" 52, because the entire purpose is to give the user the correct config, not have him manually choose it.

> we should wait until uptake of 60 is more complete until we merge this to prod.

Agreed.

**Comment by Sancus** [comment]
https://addons.thunderbird.net/en-US/thunderbird/addon/lightning/statistics/usage/applications/

You can use lightning stats for this, even though about 20% of users don't have Lightning installed for some reason, the breakdown still ends up being within a couple percentage points of all users. So we're at ~70% of users on 60.* right now.

**Comment by Sancus** [comment]
Merged, this is live now.

**Comment by mkmelin** [comment]
Great! Thx Vikram for the work on this.
